File: binance4j-core/src/test/java/com/binance4j/param/MapTest.java
package com.binance4j.param;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.Map;

import org.junit.jupiter.api.Test;

import com.binance4j.core.annotation.Param;
import com.binance4j.core.param.Params;

public class MapTest {
	@Test
	void testCustom() {
		var params = new CustomParams();
		params.foo = "string";
		params.bar = 2;

		Map<String, Object> map = params.toMap();
		assertEquals(map.get("foo"), params.foo);
		assertEquals(map.get("bar"), params.bar);
		assertTrue(map.get("timestamp") != null);
		assertTrue(map.get("recvWindow") != null);
	}

	@Test
	void testNoTimestamp() {
		Map<String, Object> map = new NoTimestampParams().toMap();
		assertTrue(!map.containsKey("timestamp"));
		assertTrue(!map.containsKey("recvWindow"));
	}

	@Param(recvWindow = false, timestamp = false)
	class CustomParams implements Params {
		public String foo;
		public int bar;
	}

	@Param(recvWindow = false, timestamp = false)
	class NoTimestampParams implements Params {
	}
}

File: binance4j-core/src/test/kotlin/com/binance4j/param/MapTest.kt
package com.binance4j.param

import com.binance4j.core.annotation.Param
import com.binance4j.core.param.Params
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class MapTest {
    @Test
    fun testCustom() {
        val params = CustomParams()
        params.foo = "string"
        params.bar = 2
        val map: Map<String, Any> = params.toMap()
        Assertions.assertEquals(map["foo"], params.foo)
        Assertions.assertEquals(map["bar"], params.bar)
        Assertions.assertNotNull(map["timestamp"])
        Assertions.assertNotNull(map["recvWindow"])
    }

    @Test
    fun testNoTimestamp() {
        val map: Map<String, Any> = NoTimestampParams().toMap()
        Assertions.assertFalse(map.containsKey("timestamp"))
        Assertions.assertFalse(map.containsKey("recvWindow"))
    }

    @Param
    internal inner class CustomParams : Params {
        var foo: String? = null
        var bar = 0
    }

    @Param(recvWindow = false, timestamp = false)
    internal inner class NoTimestampParams : Params
}